src/java/org/apache/cassandra/db/lifecycle/PartialLifecycleTransaction.java

```diff
@@ -0,0 +1,211 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.db.lifecycle;
+
+import java.util.Collection;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+import org.apache.cassandra.db.compaction.OperationType;
+import org.apache.cassandra.io.sstable.SSTable;
+import org.apache.cassandra.io.sstable.format.SSTableReader;
+import org.apache.cassandra.utils.Throwables;
+
+/// Partial lifecycle transaction. This works together with a CompositeLifecycleTransaction to allow for multiple
+/// tasks using a shared transaction to be committed or aborted together. This is used to parallelize compaction
+/// operations over the same sources. See [CompositeLifecycleTransaction] for more details.
+///
+/// This class takes care of synchronizing various operations on the shared transaction, making sure that an abort
+/// or commit signal is given exactly once (provided that this partial transaction is closed), and throwing an exception
+/// when progress is made when the transaction was already aborted by another part.
+public class PartialLifecycleTransaction implements ILifecycleTransaction
+{
+    final CompositeLifecycleTransaction composite;
+    final ILifecycleTransaction mainTransaction;
+    final AtomicBoolean committedOrAborted = new AtomicBoolean(false);
+
+    public PartialLifecycleTransaction(CompositeLifecycleTransaction composite)
+    {
+        this.composite = composite;
+        this.mainTransaction = composite.mainTransaction;
+        composite.register(this);
+    }
+
+    public void checkpoint()
+    {
+        // don't do anything, composite will checkpoint at end
+    }
+
+    private RuntimeException earlyOpenUnsupported()
+    {
+        throw new UnsupportedOperationException("PartialLifecycleTransaction does not support early opening of SSTables");
+    }
+
+    public void update(SSTableReader reader, boolean original)
+    {
+        throwIfAborted();
+        if (original)
+            throw earlyOpenUnsupported();
+
+        synchronized (mainTransaction)
+        {
+            mainTransaction.update(reader, original);
+        }
+    }
+
+    public void update(Collection<SSTableReader> readers, boolean original)
+    {
+        throwIfAborted();
+        if (original)
+            throw earlyOpenUnsupported();
+
+        synchronized (mainTransaction)
+        {
+            mainTransaction.update(readers, original);
+        }
+    }
+
+    public SSTableReader current(SSTableReader reader)
+    {
+        return mainTransaction.current(reader);
```

Comment:
Is there a reason we don't wrap this one in a `synchronized` block? In looking at how it is implemented in `LifecycleTransaction`, it seems like `current(reader)` might not be thread safe.

Reply:
Fixed